addons/l10n_vn_edi/models/account_move.py

```diff
@@ -0,0 +1,719 @@
+        """
+        self.ensure_one()
+        invoice_data = {}
+        # If the request was sent but ended up failing, there is still the possibility that the invoice was saved
+        # on their system (timeout, for example)
+        if self.l10n_vn_edi_transaction_id:
+            invoice_lookup, error_message = self._l10n_vn_edi_lookup_invoice()
+            if 'result' in invoice_lookup:
+                invoice_data = invoice_lookup['result'][0]
+            # note: We do not catch errors on this endpoint for simplicity, as it should not be required.
+        else:
+            # We do not store the transaction id on the move right away so that we can avoid the above api call.
+            # When sending for the first time, we'll get the id from the file data, which we generated earlier in the flow.
+            self.l10n_vn_edi_transaction_id = invoice_json_data['generalInvoiceInfo']['transactionUuid']
+
+        # if the above request did not return data, we can assume that the invoice has failed to be created, or was never sent
+        if not invoice_data:
+            # Send the invoice to the system
+            request_response, error_message = self._l10n_vn_edi_send_request(
+                method='POST',
+                endpoint=f'InvoiceAPI/InvoiceWS/createInvoice/{self.company_id.vat}',
+                json_data=invoice_json_data,
+            )
+
+            if error_message:
+                return [error_message]
+
+            invoice_data = request_response['result']
+
+        self.write({
+            'l10n_vn_edi_reservation_code': invoice_data['reservationCode'],
+            'l10n_vn_edi_invoice_number': invoice_data['invoiceNo'],
+            'l10n_vn_edi_state': 'sent',
+        })
+
+    def _l10n_vn_edi_cancel_invoice(self, reason, agreement_document_name, agreement_document_date):
+        """ Send a request to cancel the invoice. """
+        self.ensure_one()
+        _request_response, error_message = self._l10n_vn_edi_send_request(
+            method='POST',
+            endpoint='InvoiceAPI/InvoiceWS/cancelTransactionInvoice',
+            params={
+                'supplierTaxCode': self.company_id.vat,
+                'templateCode': self.l10n_vn_edi_invoice_symbol.invoice_template_id.name,
+                'invoiceNo': self.l10n_vn_edi_invoice_number,
+                'strIssueDate': self._l10n_vn_edi_format_date(self.l10n_vn_edi_issue_date),
+                'additionalReferenceDesc': agreement_document_name,
+                'additionalReferenceDate': self._l10n_vn_edi_format_date(agreement_document_date),
+                'reasonDelete': reason,
+            },
+        )
+        if error_message:
+            raise UserError(error_message)
+
+        self.l10n_vn_edi_state = 'cancelled'
+        self.with_context(no_new_invoice=True).message_post(
+            body=_('The invoice has been cancelled for reason: %s', reason),
+        )
+        self.button_cancel()
```

Comment:
You can't do that. First, you need to follow the state machine of the invoice's state so you need to `button_draft` first.
Secondly, in case of a locked period, it could raise an error and then, it will rollback the whole RPC call.
```suggestion
            try:
                self._check_fiscalyear_lock_date()
                self.line_ids._check_tax_lock_date()

                self.button_draft()
                self.button_cancel()
            except UserError:
                pass
```

Reply:
Then I prefer to first do the reset, and let Odoo raise.
Then if no issue, do the rpc and raise if it doesn't work.

With the try & pass we could end up with a canceled invoice on the EDI that is still posted inside Odoo.